### PR TITLE
Fix and rewrite contrast color calculation, fix project-related bugs

### PR DESCRIPTION
**Pull request by silverwind**
1. The previous color contrast calculation function was incorrect at least for the `#84b6eb` where it output low-contrast white instead of black. I've rewritten these functions now to accept hex colors and to match GitHub's calculation and to output pure white/black for maximum contrast. Before and after:
Screenshot 2024-04-02 at 01 53 46Screenshot 2024-04-02 at 01 51 30

2. Fix project-related issues:

- Expose the new `ContrastColor` function as template helper and use it for project cards, replacing the previous JS solution which eliminates a flash of wrong color on page load.
- Fix a bug where if editing a project title, the counter would get lost.
- Move `rgbToHex` function to color utils.

@HesterG fyi

### CODE REVIEW

**Comment by silverwind** [comment]
I'm not exactly sure how the current formula is called but if we eventually switch it to the APCA formula it'll be good to distinguish that the passed-in color is a background color, because that formula is not symmetric:

> The APCA formula requires the text color and the foreground color to be distinguished. Swapping them gives a somewhat different, and incorrect, result.

File: web_src/js/utils/color.test.js [hunk 1]
```diff
@@ -1,21 +1,22 @@
...
+  expect(contrastColor(162, 238, 239)).toBe(false);
+  expect(contrastColor(112, 87, 255)).toBe(true);
+  expect(contrastColor(0, 134, 114)).toBe(true);
+  expect(contrastColor(228, 230, 105)).toBe(false);
+  // expect(contrastColor(216, 118, 227)).toBe(true);
+  expect(contrastColor(255, 255, 255)).toBe(false);
+  expect(contrastColor(43, 134, 133)).toBe(true);
+  expect(contrastColor(43, 135, 134)).toBe(true);
+  // expect(contrastColor(44, 135, 134)).toBe(true);
+  // expect(contrastColor(59, 182, 179)).toBe(true);
```
Comment (delvh) [inline web_src/js/utils/color.test.js]: Why commented out?

File: modules/util/color.go [hunk 2]
```diff
@@ -4,22 +4,12 @@ package util
...
 
 // Check similar implementation in web_src/js/utils/color.js and keep synchronization
 
-// Return R, G, B values defined in reletive luminance
-func getLuminanceRGB(channel float64) float64 {
-	sRGB := channel / 255
-	if sRGB <= 0.03928 {
-		return sRGB / 12.92
-	}
-	return math.Pow((sRGB+0.055)/1.055, 2.4)
```
Comment (delvh) [inline modules/util/color.go:20]: This line was simply removed now?

File: web_src/js/utils/color.test.js (hunk 1 above)
Comment (silverwind) [reply web_src/js/utils/color.test.js]: Because it's WIP.

File: modules/util/color.go (hunk 2 above)
Comment (silverwind) [reply modules/util/color.go:20]: Yes, I think this function was causing the calculation to go wrong. As I understand it, hex colors are already in SRGB color space and need no convertion.

**Comment by silverwind** [comment]
Maybe I'll reduce this PR just to get the calculation fixed and then maybe rewrite more in another PR.

File: web_src/js/utils/color.test.js (hunk 1 above)
Comment (silverwind) [reply web_src/js/utils/color.test.js]: Fixed.

File: modules/util/color.go (hunk 2 above)
Comment (silverwind) [reply modules/util/color.go:20]: So I'm not color math expert, but I do know that the previous calculation was wrong and the current one with the removal of this function fixes it and matches GitHub, so likely it's fine to remove.

**Comment by silverwind** [comment]
Ready again. I took the oppoturnity to fix two more related bugs in projects:

1. The project column color is now calculated on backend via `ContrastColor` helper, removing a flash of wrong text color on the projects column and removing the JS that did that before.
2. When renaming a column, the JS would remove the issue counter because it set `textContent` on the wrong node. Fixed that by adding a new class name on where to set.

File: modules/util/color.go [hunk 3]
```diff
@@ -47,19 +35,26 @@ func HexToRBGColor(colorString string) (float64, float64, float64) {
...
-func UseLightTextOnBackground(r, g, b float64) bool {
-	return GetLuminance(r, g, b) < 0.453
+// Given a background color, returns a black or white foreground color that the highest
+// contrast ratio. In the future, the APCA contrast function, or CSS `contrast-color` will be better.
+// https://github.com/color-js/color.js/blob/eb7b53f7a13bb716ec8b28c7a56f052cd599acd9/src/contrast/APCA.js#L42
+func UseLightText(backgroundColor string) bool {
+	if GetRelativeLuminance(backgroundColor) < 0.453 {
+		return true
+	}
+	return false
```
Comment (silverwind) [inline modules/util/color.go]:
```suggestion
	return GetRelativeLuminance(backgroundColor) < 0.453
```

File: modules/util/color.go [hunk 4]
```diff
@@ -47,19 +35,23 @@ func HexToRBGColor(colorString string) (float64, float64, float64) {
...
+// https://github.com/color-js/color.js/blob/eb7b53f7a13bb716ec8b28c7a56f052cd599acd9/src/contrast/APCA.js#L42
+func UseLightText(backgroundColor string) bool {
+	return GetRelativeLuminance(backgroundColor) < 0.453
+}
+
+func ContrastColor(backgroundColor string) string {
+	if UseLightText(backgroundColor) {
+		return "#fff"
+	}
+	return "#000"
```
Comment (silverwind) [inline modules/util/color.go]:
```suggestion
func UseLightText(backgroundColor string) bool {
	return GetRelativeLuminance(backgroundColor) < 0.453
}

// Given a background color, returns a black or white foreground color that the highest
// contrast ratio. In the future, the APCA contrast function, or CSS `contrast-color` will be better.
// https://github.com/color-js/color.js/blob/eb7b53f7a13bb716ec8b28c7a56f052cd599acd9/src/contrast/APCA.js#L42
func ContrastColor(backgroundColor string) string {
	if UseLightText(backgroundColor) {
		return "#fff"
	}
	return "#000"
```

File: web_src/js/utils/color.js [hunk 5]
```diff
@@ -1,23 +1,21 @@
...
 
-// Reference from: https://www.w3.org/WAI/GL/wiki/Relative_luminance
-function getLuminance(r, g, b) {
-  const R = getLuminanceRGB(r);
-  const G = getLuminanceRGB(g);
-  const B = getLuminanceRGB(b);
-  return 0.2126 * R + 0.7152 * G + 0.0722 * B;
+// Given a background color, returns a black or white foreground color that the highest
+// contrast ratio. In the future, the APCA contrast function, or CSS `contrast-color` will be better.
+// https://github.com/color-js/color.js/blob/eb7b53f7a13bb716ec8b28c7a56f052cd599acd9/src/contrast/APCA.js#L42
```
Comment (silverwind) [inline web_src/js/utils/color.js]:
```suggestion
```

File: web_src/js/utils/color.js [hunk 6]
```diff
@@ -1,23 +1,18 @@
...
+function useLightText(backgroundColor) {
+  return getRelativeLuminance(backgroundColor) < 0.453 ? true : false;
 }
 
-// Reference from: https://firsching.ch/github_labels.html
-// In the future WCAG 3 APCA may be a better solution.
-// Check if text should use light color based on RGB of background
-export function useLightTextOnBackground(r, g, b) {
-  return getLuminance(r, g, b) < 0.453;
+export function contrastColor(backgroundColor) {
```
Comment (silverwind) [inline web_src/js/utils/color.js:17]:
```suggestion
// Given a background color, returns a black or white foreground color that the highest
// contrast ratio. In the future, the APCA contrast function, or CSS `contrast-color` will be better.
// https://github.com/color-js/color.js/blob/eb7b53f7a13bb716ec8b28c7a56f052cd599acd9/src/contrast/APCA.js#L42
export function contrastColor(backgroundColor) {
```

File: web_src/js/utils/color.js [hunk 7]
```diff
@@ -1,23 +1,21 @@
...
 }
 
-// Reference from: https://www.w3.org/WAI/GL/wiki/Relative_luminance
-function getLuminance(r, g, b) {
-  const R = getLuminanceRGB(r);
-  const G = getLuminanceRGB(g);
-  const B = getLuminanceRGB(b);
-  return 0.2126 * R + 0.7152 * G + 0.0722 * B;
+function useLightText(backgroundColor) {
+  return getRelativeLuminance(backgroundColor) < 0.453 ? true : false;
```
Comment (delvh) [inline web_src/js/utils/color.js]:
```suggestion
  return getRelativeLuminance(backgroundColor) < 0.453;
```

File: web_src/js/utils/color.js [hunk 8]
```diff
@@ -33,3 +31,13 @@ export const chartJsColors = resolveColors({
...
 });
+
+function hex(x) {
+  const hexDigits = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'];
+  return Number.isNaN(x) ? '00' : hexDigits[(x - x % 16) / 16] + hexDigits[x % 16];
+}
+
+export function rgbToHex(rgb) {
+  rgb = rgb.match(/^rgba?\((\d+),\s*(\d+),\s*(\d+).*\)$/);
+  return `#${hex(rgb[1])}${hex(rgb[2])}${hex(rgb[3])}`;
```
Comment (delvh) [inline web_src/js/utils/color.js]: So we simply ignore the `alpha` when it exists?

File: web_src/js/utils/color.js (hunk 7 above)
Comment (silverwind) [reply web_src/js/utils/color.js]: Good catch, there ought to be a lint rule for this somewhere.

File: web_src/js/utils/color.js (hunk 8 above)
Comment (silverwind) [reply web_src/js/utils/color.js]: I did not modify that function, it was just moved as-is from the projects file.

And yes, projects bg does not support transparency.

File: web_src/js/utils/color.js (hunk 7 above)
Comment (silverwind) [reply web_src/js/utils/color.js]: Found it, will include in next eslint update.

https://eslint.org/docs/latest/rules/no-unneeded-ternary

File: web_src/js/utils/color.js (hunk 8 above)
Comment (delvh) [reply web_src/js/utils/color.js]: But who says that this function will only be used for projects?
It does not sound like only being intended for projects…

File: web_src/js/utils/color.js (hunk 7 above)
Comment (silverwind) [reply web_src/js/utils/color.js]: Enabled it in https://github.com/go-gitea/gitea/pull/30279

**Review by delvh — APPROVED** [review]

File: web_src/js/utils/color.js (hunk 8 above)
Comment (silverwind) [reply web_src/js/utils/color.js]: Hmm let me check if we maybe can get rid of it. We have `tinycolor2` which may be able to do it.

Comment (silverwind) [reply web_src/js/utils/color.js]: Done in https://github.com/go-gitea/gitea/pull/30237/commits/4ac99163d0e966960ab3465b1046daa5af922848.

**Comment by silverwind** [comment]
Found two more bugs:

- Moving project boards sends the computed background color to backend, even for boards with unset color, resulting in dark theme sending a different one than light theme.
- If a board has no background color set, we render a empty string to the template and `ContrastColor` gives incorrect `#fff`.

**Comment by silverwind** [comment]
> Found two more bugs:
> 
>     * Moving columns sends the computed background color to backend, even for boards with unset color, resulting in dark theme sending a different value than light theme. If at alle, it should send empty string for uncolored columns.
> 
>     * If a column has no background color set, we render a empty string to the template and `ContrastColor` gives incorrect `#fff`.

Second issue is fixed. Also the divider is now in correct color in all cases:

image

**Comment by silverwind** [comment]
Second bug fixed as well in https://github.com/go-gitea/gitea/pull/30237/commits/955d63479107ec44d5946b88dc6d034095e2a360. Now it will send empty string for uncolored columns. `getComputedStyle` wasn't necessary at all and likely this was a regression from the `.css` refactoring.

File: web_src/js/components/ContextPopup.vue [hunk 9]
```diff
@@ -108,7 +102,7 @@ export default {
       <p><small>{{ issue.repository.full_name }} on {{ createdAt }}</small></p>
       <p><svg-icon :name="icon" :class="['text', color]"/> <strong>{{ issue.title }}</strong> #{{ issue.number }}</p>
       <p>{{ body }}</p>
-      <div>
+      <div class="tw-flex tw-flex-wrap tw-gap-1">
```
Comment (wxiaoguang) [inline web_src/js/components/ContextPopup.vue]: I think it could be `flex-text-block tw-flex-wrap`

**Review by wxiaoguang — APPROVED** [review]

File: web_src/js/components/ContextPopup.vue (hunk 9 above)
Comment (silverwind) [reply web_src/js/components/ContextPopup.vue]: Even better, we can share `#issue-list .flex-item-title .labels-list` style.

Comment (silverwind) [reply web_src/js/components/ContextPopup.vue]: Done in https://github.com/go-gitea/gitea/pull/30237/commits/a26de2a6752ff117b8121f8fc1cd359b698e6691, now there is a single shared `labels-list` class.

**Pull request merged**